File: android/domain/src/main/kotlin/com/bolsillo/domain/usecase/RestoreTransaction.kt
package com.bolsillo.domain.usecase

import com.bolsillo.domain.port.TransactionRepository
import javax.inject.Inject

/** FR 13 — restore from trash. Transfer legs come back together as a pair. */
class RestoreTransaction
    @Inject
    constructor(private val transactions: TransactionRepository) {
        suspend operator fun invoke(id: String) {
            val target = transactions.getById(id) ?: return
            val group = target.transferGroupId
            if (group != null) {
                transactions.restoreGroup(group)
            } else {
                transactions.restore(id)
            }
        }
    }

File: ios/Packages/Bolsillo/Sources/BolsilloDomain/UseCases/RestoreTransaction.swift
import Foundation

/// Restores a soft-deleted transaction from trash. If the transaction belongs to a
/// transfer group, both legs are restored atomically to preserve the double-entry
/// invariant (Constitution Article III — Invariants 6 & 7).
public struct RestoreTransaction: Sendable {
    private let transactions: any TransactionRepository

    public init(transactions: some TransactionRepository) {
        self.transactions = transactions
    }

    /// Restore the transaction identified by `id`.
    ///
    /// - Parameter id: The transaction to restore. If it belongs to a transfer group,
    ///   the entire group is restored.
    public func callAsFunction(id: String) async {
        if let tx = await transactions.getById(id), let groupId = tx.transferGroupId {
            // Transfer: restore the whole group atomically (Invariant 7)
            await transactions.restoreGroup(transferGroupId: groupId)
        } else {
            await transactions.restore(id: id)
        }
    }
}
